File: src/commands/signup.ts
import { CommandInteraction, GuildMember } from "discord.js";
import Command from "../structures/Command.js";
import { groups } from "../utils/Constants.js";

const choices = Object.entries(groups)
	.map(([value, { name }]) => ({ name, value }));

function capitalize(name: string) {
	return name
		.toLowerCase()
		.replace(/[ -]([a-z]){1}|(^[a-z]){1}/g, m => m.toUpperCase());
}

export default class SignupCommand extends Command {
	constructor() {
		super("inscription", {
			description: "Renseignez votre identité pour accéder au serveur",
			args: {
				prenom: { as: "string", desc: "Votre prénom" },
				nom: { as: "string", desc: "Votre nom de famille" },
				groupe: { as: "choice", desc: "Votre classe (ou catégorie)", choices }
			}
		});
	}

	run(interaction: CommandInteraction) {
		const firstname = interaction.options.get("prenom").value as string;
		const lastname = interaction.options.get("nom").value as string;
		const group = interaction.options.get("groupe").value as string;

		const member = interaction.member as GuildMember;

		Promise.all([
			member.setNickname(capitalize(`${firstname} ${lastname}`)),
			member.roles.add(groups[group].roles)
		]).then(() => {
			interaction.reply({
				content: "Vous avez été identifié avec succès !",
				ephemeral: true
			});
		}).catch(err => {
			console.error(err);
			interaction.reply({
				content: "Quelque chose s'est mal passé ! Veuillez contacter un administrateur du serveur.",
				ephemeral: true
			});
		});
	}
}

File: src/commands/signup.rs
use poise::serenity_prelude::CacheHttp;

use crate::{
    config::{Group, CONFIG},
    Context, Error,
};

/// Renseignez votre identité pour accéder au serveur
#[poise::command(slash_command)]
pub async fn inscription(
    ctx: Context<'_>,
    #[description = "Votre prénom"] prenom: String,
    #[description = "Votre nom"] nom: String,
    #[description = "Votre classe (ou catégorie)"] groupe: Group,
) -> Result<(), Error> {
    let member = ctx.author_member().await.unwrap();

    // checks that the user is not already registered
    if CONFIG.roles_contains(member.roles.as_slice()) {
        ctx.send(|reply| {
            reply
                .content("Il semblerait que vous soyez déjà inscrit. Pour changer de classe, utilisez la commande /changer-classe.")
                .ephemeral(true)
        }).await?;

        return Ok(());
    }

    member
        .edit(ctx.http(), |edit| {
            edit.nickname(format!("{} {}", prenom, nom))
        })
        .await?
        .add_roles(ctx.http(), groupe.to_roleids().as_slice())
        .await?;

    ctx.send(|reply| {
        reply
            .content("Vous avez été identifié avec succès !")
            .ephemeral(true)
    })
    .await?;

    Ok(())
}
